src/docs/binding-classes.md

```diff
@@ -0,0 +1,397 @@
+
+Swift initializers can fail - this is something that doesn't have an exact analog in C#. There are two ways to handle this:
+- make the corresponding C# constructor throw
+- write a factory method that returns `BoundType?`
+
+## ABI differences
+- Instance methods get called with the `self` register pointing to the instance.
+- Class methods get called with the `self` register pointing to the type metadata.
+- Init methods get called with the type metadata in the first argument.
+- Methods that throw return the error in the error register.
+- Virtual methods in swift get called via a vtable. Since the ordering of the vtable is undefined, methods can't be called using the vtable from C#. In addition, the vtable is consider fragile and may change between versions of any given library. Fortunately, with enable-library-evolution set, the compiler will inject a dispatch thunk for each virtual method.
+
+
+## Runtime Differences
+Swift classes are reference counted rather than garbage collected. This can create some interesting conditions when an object from Swift surfaced in C# gets garbage collected, but the object may still be live in Swift, so it's necessary to manage this. The way this is handled in Binding Tools for Swift is to add it to an object registry which is a `Dictionary&lt;NativeHandle, GCHandle&gt;` The native handle is the Swift object instance and the GC Handle is made from the corresponding C# instance. When the pair is added to the registry, the registry also takes a weak reference to the Swift
+object in order to insure that it doesn't away if it gets disposed.
+
+Naturally, when the C# object gets disposed, the registry removes and does a weak release of the object.
+
+## Idiomatic Differences
+There are very few idiomatic differences between Swift classes and C# classes.
+
+## Accessibility
+This is where the true challenge lies.
+Let's break Swift classes into two varieties: virtual and non-virtual. Non-virtual are easy because the implementation will be a `sealed` C# class for which each method would be a p/invoke into the Swift implementation.
+
+Virtual classes are another matter.
+
+The goal would be to have subclasses of the bound Swift class to be able to operate in Swift.
+
+To do this, you need to write Swift code that overrides all the virtual methods in the class and delegates them out to C#. In Binding Tools for Swift, this is was done by created a simulated vtable in Swift and populating it from C#.
+
+```swift
+// original swift class
+open class YourAge {
+    public init () { }
+    open func getAge() -> Int {
+        return 7
+    }
+}
+
+// simulated vtable:
+fileprivate struct YourAge_VTable {
```

Comment:
Why do we need the vtable? Can't we just expose public members?

Reply:
Recall that purpose of this is to make it so that Swift can call into C# for the actual implementation of the members. For Swift to get the function pointer, C# must first call into Swift to set them. For n virtual members, we could either (1) make n calls to set the function pointers, (2) 1 call with n arguments, or (3) 1 call taking 1 argument (a vtable). While 2 and 3 are more or less equivalent, it won't work in the case of generic classes with virtual members, for which each class has to keep a dictionary of bound generic types mapped to functions implementing the virtual methods.
Because of efficiency and uniformity, having a vtable (per unique generic binding) makes the most sense.

Reply:
The challenge is that Swift needs to know a priori how to call C# from Swift method overrides. There's effectively a circular dependency here. Are you suggesting that C# exports symbols that Swift will have access to? I can see how this might work, but how is it in an improvement? Most of the actual work will be in marshaling since Swift really does not like passing reference counted types to C function pointers.

Reply:
There's two scenarios I'm addressing here:

1. Passing an instance of a C# class that inherits from a Swift base class as a reference to the base class:

In this case, the projection tooling should generate the type metadata pointer in the Swift object wrapper for the C# object such that the type metadata and vtables are set up with pointers to the C# overrides. These functions will have the **Swift** calling convention, not the C calling convention.

Therefore, when the Swift compiler generates code to call methods on the vtables of (possibly-derived) instances of the Swift base class, the metadata is correctly constructed such that the call goes to the `UnmanagedCallersOnly` wrapper with the Swift calling convention that the projection tooling generates.

This would remove any requirements to emit Swift code for this scenario, as the projection could construct the code into the correct structure from the start.

This would require us to know how the vtable is sorted. However, the vtable must be sorted in some sort of stable fashion (or have a mechanism to look up members by name) to support the `enable-library-evolution` scenario. We can work with the folks at Apple to document the existing behavior as it is set in stone until they explicitly break ABI.

2. Exposing a C# class to be instantiated from Swift (not currently in scope)

In this case, when using a toolchain like NativeAOT, the correct symbols (pointing to the data produced in scenario 1) are exported as per the Swift docs so Swift code that links to the NativeAOT shared library correctly discovers all of the expected symbols to link and the Swift compiler can treat the NativeAOT binary as a Swift binary (as it exposed the same symbols as if the library had been authored in Swift).

Reply:
OK - great. Thanks for explaining what you're thinking of. Scenario 1 is effectively impossible. `enable-library-evolution` dictates that the order of the vtable is undefined. (see the doc I linked from protocols). Resiliency is managed through dispatch thunks. The relevant paragraph is:
> The key takeaway here is that method dispatch to Swift-native methods on a resilient class is performed by calling a dispatch thunk; as with protocols, this allows methods on the class to be re-ordered and new methods added without disturbing callers. This mechanism also allows the superclass to add or remove methods, without disturbing subclasses.

Now, is it possible for us to write code that will generate a Swift type that will dispatch into C#? Yes, and in order to avoid having to implement our own code to generate type metadata etc, we let the Swift compiler do all of that for us.

But - here is an interesting thought for an optimization to that process - what if we wrote a Swift implementation in a way that had really obvious code for each of the method implementations and then rewrote the compiled code at link time or runtime to do go where we wanted.  So think about it like this - suppose I have the following code and wrapper:
```swift
open class OverrideMe { // assume that this came from someone else's library
    public init () { }
    open func multiName(text: String) -> String {
        return text + text
    }
}

open class OurWrapper : OverrideMe { // we generate this
    var gcHandle: OpaquePointer
    public init(gcHandle: OpaquePointer) {
        super()
        self.gcHandle = gcHandle
    }
    public func multiNameSuper(text: String, handle: OpaquePointer) -> String
    {
         return super.multiName(text: text)
    }
    open override multiName(text: String) -> String {
        return multiNameSuper(text: text, handle: gcHandle) // this is the important line
    }
}
```

So now imagine that we root through the binary, find the override and change the call site from `multiNameSuper` to a reference to the C# `[UnmanagedCallersOnly]` callback function. What this does is turn the responsibility of setting the callback pointer to a post-aot step.